bin/coqtop.ml

```diff
@@ -0,0 +1,141 @@
+          from the $(b,dune) files, but does not require maintaining a
+          $(b,_CoqProject) file.|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "top" ~doc ~man
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coqtop =
+    let doc = "Run the given toplevel command instead of the default." in
+    Arg.(value & opt string "coqtop" & info [ "toplevel" ] ~docv:"CMD" ~doc)
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        (* Checks on the argument file. *)
+        let source = Common.prefix_target common coq_file_arg in
+        if not (Path.is_file (Path.of_string source)) then
+          User_error.raise [ Pp.textf "cannot find file: %s" coq_file_arg ];
+        let coq_file_arg =
+          let cwd = Path.external_ (Path.External.initial_cwd) in
+          let file = Path.of_filename_relative_to_initial_cwd coq_file_arg in
+          let cwd = Path.to_string cwd in
+          let file = Path.to_string file in
+          match String.drop_prefix ~prefix:(cwd ^ "/") file with
+          | None -> coq_file_arg
+          | Some s -> s
+        in
+        let coq_file_build =
+          let p = Common.prefix_target common coq_file_arg in
+          Path.Build.relative context.build_dir p
+        in
+        let dir =
+          let dir = Filename.dirname coq_file_arg in
+          let p = Common.prefix_target common dir in
+          Path.Build.relative context.build_dir p
+        in
+        let* (coqtop, args) =
+          let open Memo.Build.O in
+          Build_system.run_exn (fun () ->
+              let* dc = Dune_rules.Dir_contents.get sctx ~dir in
+              let* coq_src = Dune_rules.Dir_contents.coq dc in
+              let (lib_name, coq_module) =
+                let source = coq_file_build in
+                match Dune_rules.Coq_sources.find_module ~source coq_src with
+                | Some m -> m
+                | None -> User_error.raise
+                            [ Pp.textf "cannot find file: %s" coq_file_arg ]
+              in
+              let* _ : unit * Dep.Fact.t Dep.Map.t =
+                let deps =
+                  Dune_rules.Coq_rules.deps_of ~dir ~boot_type:No_boot coq_module
+                in
+                Action_builder.run deps Eager
+              in
+
+              let* conf = Dune_rules.Dune_load.load () in
+              let* dune_files =
```

Comment:
This is all needed to get the theory stanza with a coq module, right? Unfortunately, this isn't quite the right way to do this. Luckily, we have code that does the same thing for OCaml, so you should be able to adapt it to coq easily. Take a look at the implementation of `Ml_sources.lookup_module`.

Reply:
Yes, I wrote this to extract a theory stanza given a Coq module. I've had a look to `Ml_sources.lookup_module`, and I'm not quite sure I understand what you want me to do. The `Ml_sources.lookup_module` returns a `Dune_file.Buildable.t`, which does not seem to contain any Coq-related things. So, my guess is that we want something of type `Coq_stanza.Buildable.t` instead. However, I don't see how I can get a `Coq_stanza.Theory.t` from that. Could you elaborate a bit?

Side question: if I am right that `Dune_file.Buildable.t` does not give any information about Coq theories, why is it contained in a generic-sounding module like `Dune_file`?

Reply:
My first response was definitely low on the details. Let me try to explain things a little more and hopefully it should be clear. 

Conceptually, `Ml_sources.t` is a map from buildable OCaml stanzas (libraries, executables) to all relevant OCaml modules. By analogy, `Coq_sources.t` is a map from Coq stanzas (theories, extractions) to all relevant Coq modules. Note that that keyword "relevant" here is determined by how `(include_subdirs ..)` is interpreted, but it's not important for what we're working on.

What's important is the main difference between `Ml_sources.t` and `Coq_sources.t`. If both are maps from stanzas to modules, then `Ml_sources.t` additionally constructs the reverse map (modules to stanzas) as well. `Coq_sources.t` does not construct such a reverse map, and therefore does not offer a function analogous to `lookup_module`. My suggestion was to complete `Coq_sources.t` and construct the two way map.

For coq, the signature would be something like:

```
val lookup_module : Coq_module.Name.t -> [`Theory of Coq_stanzy.Theory.t | `Extraction of Coq_stanza.Extraction.t ]
```

> Side question: if I am right that Dune_file.Buildable.t does not give any information about Coq theories, why is it contained in a generic-sounding module like Dune_file?

Dune is a very big project at this point, and not everything is well organized. Dune_file is just a grab bag of things related to parsing stanzas. It should be split eventually.

Reply:
I see, thanks for the detailed explanation!

Reply:
Fixed in 9d1fe8c67d5a8e63e0d64a0e624184d0f27e8610, though the `lookup_module` function has a slightly different type from what you suggested.

Reply:
Looks fine. So this feature isn't going to work for loading `.v` files used for extraction right?

Reply:
I'm not sure I understand: in my mind, extraction is about creating `.ml` files from `.v` files, and I don't think these `.v` files that are the source of extraction need to be treated any differently from other `.v` files. I might be wrong though, @ejgallego what do you think?

Reply:
> and I don't think these .v files that are the source of extraction need to be treated any differently from other .v files

I think that's my point, no? You are excluding them from the reverse map and making it impossible to load them through coqtop. Perhaps that's desirable, but it does mean that they are treated a bit differently.

Reply:
Are you thinking of the case where you have a `.v` file that is not part of a theory, but only used for extraction?

Reply:
Yeah, exactly.

Reply:
OK, I have never used extraction, so I did not know you could do that. This should now be supported with ea2b13fa9b7eb880f30b5e5522ea65bc94427bb8.